File: linked_list/c/ex_1.c
/**
 * Remove Dups: Write code to remove duplicates from an unsorted linked list.
 * FOLLOW UP: How would you solve this problems if a temporary buffer is not
 * allowed?
 */

#include <stdio.h>
#include "linked_list.h"
#include "../../lib/c/uthash/uthash.h"

void remove_dups_buffered(list *list_p);
void remove_dups_unbuffered(list *list_p);

int main() {
    list *test_list = create_list();
    insert(test_list, 1);
    insert(test_list, 1);
    insert(test_list, 1);
    insert(test_list, 4);
    insert(test_list, 5);
    insert(test_list, 7);
    insert(test_list, 7);
    insert(test_list, 2);
    insert(test_list, 3);
    insert(test_list, 3);
    insert(test_list, 3);

    remove_dups_unbuffered(test_list);
    print_list(test_list);
}

// O(n) time complexity and O(n) space complexity
void remove_dups_buffered(list *list_p) {
    node *node_table = NULL;
    node *found = NULL;

    node *current = list_p->head;
    // Head can't be a duplicate, so add it to the hash table
    HASH_ADD_INT(node_table, data, current);

    while (current->next != NULL) {
        HASH_FIND_INT(node_table, &(current->next->data), found);
        if (found == NULL) {
            HASH_ADD_INT(node_table, data, current->next);
            current = current->next;
        } else {
            node *dup = current->next;
            current->next = dup->next;
            free(dup);
        }
    }
}

// O(n^2) time complexity and O(1) space complexity
void remove_dups_unbuffered(list *list_p) {
    node *current = list_p->head;
    
    while (current != NULL) {
        node *runner = current;
        while (runner->next != NULL) {
            if (current->data == runner->next->data) {
                // Duplicate
                node *dup = runner->next;
                runner->next = dup->next;
                free(dup);
            } else {
                runner = runner->next;
            }
        }
        current = current->next;
    }
}

File: go/linked_lists/ctci/ex1.go
package ctci

// 2.1 Remove Dups: Write code to remove duplicates from an unsorted linked
// list.
//
// FOLLOW UP: How would you solve this problem if a temporary buffer is not
// allowed?
//
// Assumptions:
//	* Confirm whether list is singly or doubly linked.
//  * For similar questions, confirm whether list is sorted.

// DedupSingle checks for duplicates in a singly linked list by storing the data
// from each seen node in a set and checking subsequent nodes against those that
// have already been seen.
//
// Time: O(n)
// Space: O(n)
func DedupSingle(head *SNode) {
	seen := make(map[int]bool)
	var prev *SNode
	for cur := head; cur != nil; cur = cur.Next {
		if seen[cur.Data] {
			prev.Next = cur.Next
		} else {
			seen[cur.Data] = true
			prev = cur // as head is never a dup, prev is always assigned before it is accessed
		}
	}
}

// DedupDouble checks for duplicates in a doubly linked list by storing the data
// from each seen node in a set and checking subsequent nodes against those that
// have already been seen.
//
// Time: O(n)
// Space: O(n)
func DedupDouble(head *DNode) {
	seen := make(map[int]bool)
	var prev *DNode
	for cur := head; cur != nil; cur = cur.Next {
		if seen[cur.Data] {
			prev.Next = cur.Next
			if cur.Next != nil {
				cur.Next.Prev = prev
			}
		} else {
			seen[cur.Data] = true
			prev = cur
		}
	}
}

// DedupSingleNoBuffer checks for duplicates by visiting each node in the
// singly linked list in turn and checking the rest of the list for matches.
//
// Time: O(n^2)
// Space: O(1)
func DedupSingleNoBuffer(head *SNode) {
	for uniq := head; uniq != nil; uniq = uniq.Next {
		for cur := uniq; cur.Next != nil; cur = cur.Next {
			if cur.Next.Data == uniq.Data {
				cur.Next = cur.Next.Next
			}
		}
	}
}

// DedupDoubleNoBuffer checks for duplicates by visiting each node in the
// doubly linked list in turn and checking the rest of the list for matches.
//
// Time: O(n^2)
// Space: O(1)
func DedupDoubleNoBuffer(head *DNode) {
	for uniq := head; uniq != nil; uniq = uniq.Next {
		for cur := uniq; cur.Next != nil; cur = cur.Next {
			if cur.Next.Data == uniq.Data {
				cur.Next = cur.Next.Next
				if cur.Next != nil {
					cur.Next.Prev = cur
				}
			}
		}
	}
}
